pyanaconda/install_tasks.py

```diff
@@ -0,0 +1,400 @@
+    """
+
+    def __init__(self, name, task):
+        self._name = name
+        self._callable = task
+        self._done = False
+        self._running = False
+        self._lock = RLock()
+        self._parent = None
+        self._start_timestamp = None
+        self._done_timestamp = None
+        self.started = Signal()
+        self.completed = Signal()
+
+    @property
+    def name(self):
+        """Task name.
+
+        :returns: task name
+        :rtype: str
+        """
+
+        return self._name
+
+    @property
+    def running(self):
+        """Reports if the task is currently running.
+
+        :reports: if the task is running
+        :rtype: bool
+        """
+        with self._lock:
+            return self._running
+
+    @property
+    def elapsed_time(self):
+        """Elapsed time since the task has been started in miliseconds."""
+        with self._lock:
+            if self._start_timestamp:
+                if self._done_timestamp:
+                    return self._done_timestamp - self._start_timestamp
+                else:
+                    return time.time() - self._start_timestamp
+            else:
+                return None
+
+    @property
+    def done(self):
+        """Reports if the task has finished processing.
+
+        :reports: if the task is done
+        :rtype: bool
+        """
+        with self._lock:
+            return self._done
+
+    @property
+    def summary(self):
+        return "Task: %s" % self.name
+
+    @property
+    def parent(self):
+        return self._parent
+
+    @parent.setter
+    def parent(self, parent_item):
+        self._parent = parent_item
+        # Connect own signals "up" to the parent
+        self.started.connect(self._parent.task_started)
+        self.completed.connect(self._parent.task_completed)
+
+    def start(self):
+        do_start = False
+        with self._lock:
+            # the task can only be started once
+            if self.running or self.done:
+                if self.running:
+                    # attempt to start a task that is already running
+                    log.error("Can't start task %s - already done.")
+                else:
+                    # attempt to start a task that an already finished task
+                    log.error("Can't start task %s - already done.")
+            else:
+                do_start = True
+                self._running = True
+                self._start_timestamp = time.time()
+
+        if do_start:
+            # trigger the "started" signal
+            self.started(self)
+            # start the task
+            self._callable()
+            # trigger the "completed" signal
+            self.completed(self)
+            # the task should be done, set the task state accordingly
+            with self._lock:
+                self._running = False
+                self._done = True
+                self._done_timestamp = time.time()
+
+def item_counter(item_count):
```

Comment:
Yeah, it's not really task/task queue specific & could be useful elsewhere, so moving to iutil. :)